### ISSUE TITLE
Username uniqueness should be enforced at the database level

### ISSUE THREAD

**Issue by dcinzona** [OrchardCMS/Orchard#5163]
Unless I'm not thinking clearly, there should never be a scenario when there are multiple users with the same username.  All usernames should be unique and this should be required and enforced on the DB level if possible.

Right now, it is only enforced at the UserService level.  We have seen some weird behavior where we have duplicate user accounts created in the Orchard DB, where the User ID is only 1 number apart - making me think there's a race condition somewhere (in our code).

That being said, we have a custom membership service that uses the UserService to create and manage user accounts - and are syncing those accounts with an external repository for future SSO support.

**Comment by pszmyd** [comment]
The solution would be to perform checks on application level and remove the race condition.
We're not adding DB-level UNIQUE constraints (as well as NOT NULL) on purpose. NHibernate controls when and how data is flushed to the database. With constraints on this could lead to undesired behavior.

**Comment by dcinzona** [comment]
Under every single other circumstance, I agree.  I don't agree, however, when it comes to something as application critical as unique user names and identity.

While this can be enforced at the application layer, it should be so that even if it isn't enforced at the database layer, that there would be absolutely no way to create two users with the same username.

The race condition occurs because we are creating a mirrored user account in an external database and piggy backing off of that for authentication, so that we can implement SSO.  The other part of this is that the UserPart has an additional UserProfilePart associated with it, where the UserProfilePart includes first and last name.

My hunch is that the user record is being created twice in this case, but I'm not sure why... and it's incredibly difficult to reproduce (i.e. doesn't happen every time).

Out of 7500 user accounts, we have 13 that are duplicates.  That's how rare of an occurrence this is. 

I will look into it on the coding end some more, but thought that enabling a uniqueness constraint could help point out when and where this issue occurs.  In fact, I may try setting that constraint locally to see if I can get it squared away.

**Comment by dcinzona** [comment]
Closing this - as changing the schema would never be feasible today.

**Comment by dcinzona** [comment]
Actually, reopening this.
Can others verify that they have duplicate records in their UserPart table?
```SQL
SELECT TOP 1000 [Id]
      ,[UserName]
      ,[Email]
      ,[NormalizedUserName]
      ,[Password]
      ,[PasswordFormat]
      ,[HashAlgorithm]
      ,[PasswordSalt]
      ,[RegistrationStatus]
      ,[EmailStatus]
      ,[EmailChallengeToken]
  FROM [dbo].[Orchard_Users_UserPartRecord] as b

  where b.UserName in (
  Select UserName from [dbo].[Orchard_Users_UserPartRecord] as a
  group by UserName having count(UserName) > 1
  )

  order by username,id ```

**Comment by ASADeveloper** [comment]
I agree with dcinzona. When we install 2 instances of Orchard on 2 different servers (like Farm scenario), and point them to the same database there is no way to enforce uniqueness of username reliably on application level. So Sometimes the uniqueness of data should be enforced on the storage level.

**Comment by dcinzona** [comment]
I think I figured out why there's a problem with enforcing uniqueness.  Orchard, by design, doesn't perform hard deletes.  This means if you delete a user, the user still exists in the table.  Therefor, they wouldn't be able to enforce the uniqueness constraint lest old usernames also be unavailable. 

My recommendation would be very difficult to implement, and that is to perform hard deletes on security related content, such as user accounts.  I've noticed that the database retains users AND their old passwords.  I understand that this allows for the ability to restore old or deleted accounts and retain the password - but from a security perspective, these should be completely deleted. 

In fact, maybe ONLY the users module should perform hard deletes. 

I'm not sure how this would affect the Audit module in 1.9....
